_posts/2022-05-23-fine-grained-access-control-for-multitenancy.markdown

```diff
@@ -0,0 +1,369 @@
+---
+layout: post
+title: Fine Grained Access Control for Multi-tenancy
+description: A how to guide for architecting multitenant backend applications
+date: 2022-05-23
+author: Leon Ticharwa
+image: img/blog/multitenancy/multitenant.png
+tags: ['multitenant', 'typescript', 'tutorial', 'dynamodb', 'lambda', 'api gateway', 'authorization', 'authentication', 'apiGateway', 'jwt', 'json web token', 'cognito']
+---
+
+## What is Multi-tenancy
+
+Multi-tenancy is a software architecture that allows multiple users to access resources that reside in a shared computing environment. This architecture grants groups of users or individual users access to resources as defined by access boundaries put in place via the use of resource policies, IAM policies or some other similar mechanism. This concept is analogous to an apartment building where each tenant is given keys that only unlock their apartment door and other rooms within the apartment.
+
+Architecting multi-tenant cloud native software can be quite challenging due to the broad scope of technical know-how required to execute such an undertaking.
+What's even more challenging is finding cohesive guides/tutorials that take a top down approach to the somewhat elusive topic of multitenancy. This article is an attempt to produce a distilled "how-to" guide for those who may find themselves bogged down while building multi-tenant applications. You can clone the repo containing the reference code from here [reference code here](https://github.com/MechanicalRock/Multitenancy-AuthorizationAuthentication).
+
+## Some need to knows (Authorization and Authentication)
+
+For new developers, it is not immediately obvious that there is a difference between the terms Authorization and Authentication. Although related, these terms actually refer to two different concepts. Authentication verifies a user's identity whereas authorization verifies what a user is allowed to access once they have been authenticated. In short, authentication is about who is allowed in and authorization is about what they are allowed to access once they are in.
+
+In this write up I'll demonstrate how one might go about using Cognito to develop the authentication and authorization layer in a contextualised scenario.
+
+### JSON Web Tokens
+
+JWT, JSON Web Token, is an open standard that is widely used to securely share authentication information (claims) between client and server. The standard is defined in the RFC7519 specification developed by the Internet Engineering Taskforce (IETF).
```

Comment:
Can't make as a suggestion for some reason...

JSON Web Token (JWT)

Abbreviations under style guide